sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Cache.java

```diff
@@ -0,0 +1,259 @@
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+
+/** Transforms for reading and writing request/response associations to a cache. */
+class Cache {
+
+  /**
+   * Instantiates a {@link Call} {@link PTransform} that reads {@link RequestT} {@link ResponseT}
+   * associations from a cache. The {@link KV} value is null when no association exists. This method
+   * does not enforce {@link Coder#verifyDeterministic} and defers to the user to determine whether
+   * to enforce this given the cache implementation.
+   */
+  static <
+          @NonNull RequestT,
+          @Nullable ResponseT,
+          CallerSetupTeardownT extends
+              @NonNull Caller<@NonNull RequestT, KV<@NonNull RequestT, @Nullable ResponseT>>
+                  & SetupTeardown>
+      PTransform<
+              @NonNull PCollection<@NonNull RequestT>,
+              Call.@NonNull Result<KV<@NonNull RequestT, @Nullable ResponseT>>>
+          read(
+              @NonNull CallerSetupTeardownT implementsCallerSetupTeardown,
+              @NonNull Coder<@NonNull RequestT> requestTCoder,
+              @NonNull Coder<@Nullable ResponseT> responseTCoder) {
+    return Call.ofCallerAndSetupTeardown(
+        implementsCallerSetupTeardown, KvCoder.of(requestTCoder, responseTCoder));
+  }
+
+  /**
+   * Instantiates a {@link Call} {@link PTransform}, calling {@link #read} with a {@link Caller}
+   * that employs a redis client.
+   *
+   * <p>This method requires both the {@link RequestT} and {@link ResponseT}s' {@link
+   * Coder#verifyDeterministic}. Otherwise, it throws a {@link NonDeterministicException}.
+   *
+   * <p><a href="https://redis.io">Redis</a> is designed for multiple workloads, simultaneously
+   * reading and writing to a shared instance. See <a
+   * href="https://redis.io/docs/get-started/faq/">Redis FAQ</a> for more information on important
+   * considerations when using this method to achieve cache reads.
+   */
+  static <@NonNull RequestT, @Nullable ResponseT>
+      PTransform<
+              @NonNull PCollection<@NonNull RequestT>,
+              Call.@NonNull Result<KV<@NonNull RequestT, @Nullable ResponseT>>>
+          readUsingRedis(
+              @NonNull RedisClient client,
+              @NonNull Coder<@NonNull RequestT> requestTCoder,
+              @NonNull Coder<@Nullable ResponseT> responseTCoder)
+              throws NonDeterministicException {
+    return read(
+        new UsingRedis<>(requestTCoder, responseTCoder, client).read(),
+        requestTCoder,
+        responseTCoder);
+  }
+
+  /**
+   * Write a {@link RequestT} {@link ResponseT} association to a cache. This method does not enforce
+   * {@link Coder#verifyDeterministic} and defers to the user to determine whether to enforce this
+   * given the cache implementation.
+   */
+  static <
+          @NonNull RequestT,
```

Comment:
I removed `@NonNull` and `./gradlew :sdks:java:io:rrio:build` still runs. Do I understand true when checkerframework enabled these template parameters are NonNull by default?